### PR TITLE
[NET-562] Persistent Keep Alive from node to host

### PR DESCRIPTION
**Pull request by gabrielseibel1**
## Describe your changes

Move PKA from node to host level

## Provide testing steps

Described in ticket

### CODE REVIEW

**Comment by Aceix** [comment]
Will definitely need a UI change as well

File: logic/hosts.go [hunk 1]
```diff
@@ -6,18 +6,17 @@ import (
...
-	"sync"
-
 	"github.com/devilcove/httpclient"
 	"github.com/google/uuid"
 	"github.com/gravitl/netmaker/database"
 	"github.com/gravitl/netmaker/logger"
 	"github.com/gravitl/netmaker/models"
 	"github.com/gravitl/netmaker/servercfg"
 	"golang.org/x/crypto/bcrypt"
+	"net/http"
```
Comment (mattkasun) [inline logic/hosts.go:16]: re-order imports

**Review by mattkasun — CHANGES_REQUESTED** [review]
when a server is updated to this version; how are the host and node models updated.  The new node model will just ignore the keepalive in the database, but how will the host.Keepalive value be set?

**Review by Aceix — CHANGES_REQUESTED** [review]

File: logic/wireguard.go [hunk 2]
```diff
@@ -6,15 +6,18 @@ import (
...
 	if newNode.Address.String() != currentNode.Address.String() ||
 		newNode.Address6.String() != currentNode.Address6.String() ||
 		newNode.IsEgressGateway != currentNode.IsEgressGateway ||
 		newNode.IsIngressGateway != currentNode.IsIngressGateway ||
 		newNode.IsRelay != currentNode.IsRelay ||
-		newNode.PersistentKeepalive != currentNode.PersistentKeepalive ||
 		newNode.DNSOn != currentNode.DNSOn ||
-		newNode.Connected != currentNode.Connected {
+		newNode.Connected != currentNode.Connected ||
+		newHost.PersistentKeepalive != currentHost.PersistentKeepalive {
```
Comment (abhishek9686) [inline logic/wireguard.go:20]: this check here is not needed, since it is moved to host level

File: models/api_host.go [hunk 3]
```diff
@@ -94,6 +97,10 @@ func (a *ApiHost) ConvertAPIHostToNMHost(currentHost *Host) *Host {
 	h.IsDefault = a.IsDefault
 	h.NatType = currentHost.NatType
 	h.TurnEndpoint = currentHost.TurnEndpoint
-
+	if h.PersistentKeepalive == 0 {
+		h.PersistentKeepalive = time.Duration(a.PersistentKeepalive)
+	} else {
+		h.PersistentKeepalive = currentHost.PersistentKeepalive
+	}
```
Comment (abhishek9686) [inline models/api_host.go:104]: h.PersistentKeepalive = a.PersistentKeepalive

**Review by abhishek9686 — CHANGES_REQUESTED** [review]

File: models/api_host.go (hunk 3 above)
Comment (gabrielseibel1) [reply models/api_host.go:104]: https://github.com/gravitl/netmaker/pull/2604/commits/c02eb3a8a7cc7b702bfc9959b2b5cbc80e2ad99d

**Comment by gabrielseibel1** [comment]
> when a server is updated to this version; how are the host and node models updated. The new node model will just ignore the keepalive in the database, but how will the host.Keepalive value be set?

@mattkasun I've implemented logic for that in https://github.com/gravitl/netmaker/pull/2604/commits/f67bff1ad506a97ad91c1b46dc92600c447f7175 (https://github.com/gravitl/netmaker/pull/2604/commits/f67bff1ad506a97ad91c1b46dc92600c447f7175#diff-24c3832e2410ce37a7dfb88d3b4bf45a0506ce708774402fecddb5e96f9ff8d8R169) and in https://github.com/gravitl/netmaker/pull/2604/commits/10c34168d61084ce5568c30699a467d819190884, is that sufficient for your concern?

File: swagger.yml [hunk 4]
```diff
@@ -1136,6 +1136,8 @@ components:
                     type: string
                 TurnEndpoint:
                     type: string
+                PersistentKeepAlive:
```
Comment (mattkasun) [inline swagger.yml:1139]: swagger.yml is a generated file - should not be updated manually

**Review by mattkasun — CHANGES_REQUESTED** [review]

File: swagger.yml (hunk 4 above)
Comment (gabrielseibel1) [reply swagger.yml]: https://github.com/gravitl/netmaker/pull/2604/commits/5ca35fea91a7af804b7eb87262360b0a435f6406

File: controllers/enrollmentkeys.go [hunk 5]
```diff
@@ -155,7 +165,10 @@ func handleHostRegister(w http.ResponseWriter, r *http.Request) {
 		logic.ReturnErrorResponse(w, r, logic.FormatError(err, "internal"))
 		return
 	}
+	// set defaults
+	newHost.PersistentKeepalive = models.DefaultPersistentKeepAlive
```
Comment (abhishek9686) [inline controllers/enrollmentkeys.go:169]: only set this for a new host, check `hostexists` check below

File: logic/peers.go [hunk 6]
```diff
@@ -122,7 +122,7 @@ func GetPeerUpdateForHost(network string, host *models.Host, allNodes []models.N
 			}
 			peerConfig := wgtypes.PeerConfig{
 				PublicKey:                   peerHost.PublicKey,
-				PersistentKeepaliveInterval: &peer.PersistentKeepalive,
+				PersistentKeepaliveInterval: &peerHost.PersistentKeepalive, // TODO remove?
 				ReplaceAllowedIPs:           true,
```
Comment (abhishek9686) [inline logic/peers.go:126]: remove the TODO comment, it's necessary to set PKA  there

File: models/api_host.go (hunk 3 above)
Comment (abhishek9686) [reply models/api_host.go:104]: the above func is called for host update from UI so, there is no need to check h.PersistentKeepalive == 0, just set h.PersistentKeepalive = a.PersistentKeepalive

**Review by abhishek9686 — CHANGES_REQUESTED** [review]

File: controllers/enrollmentkeys.go (hunk 5 above)
Comment (gabrielseibel1) [reply controllers/enrollmentkeys.go]: https://github.com/gravitl/netmaker/pull/2604/commits/fcc150aad648ae394cf471046368b02416be1a97

File: logic/peers.go (hunk 6 above)
Comment (gabrielseibel1) [reply logic/peers.go:126]: https://github.com/gravitl/netmaker/pull/2604/commits/84ccc5828f17ba740f53c567158f4f089f588db8

File: models/api_host.go (hunk 3 above)
Comment (gabrielseibel1) [reply models/api_host.go]: https://github.com/gravitl/netmaker/pull/2604/commits/a51ddebde7885c4150be31c33bb484cca518efa2

**Review by abhishek9686 — APPROVED** [review]

**Comment by abhishek9686** [comment]
@gabrielseibel1  resolve conflicts

**Review by mattkasun — APPROVED** [review]

File: models/api_host.go [hunk 7]
```diff
@@ -94,6 +97,6 @@ func (a *ApiHost) ConvertAPIHostToNMHost(currentHost *Host) *Host {
 	h.IsDefault = a.IsDefault
 	h.NatType = currentHost.NatType
 	h.TurnEndpoint = currentHost.TurnEndpoint
-
+	h.PersistentKeepalive = time.Duration(a.PersistentKeepalive)
```
Comment (Aceix) [inline models/api_host.go:100]: @gabrielseibel1 i think this is the reason for the nanoseconds. should be `a.PersistentKeepalive * time.Second`

File: models/api_host.go [hunk 8]
```diff
@@ -57,6 +59,7 @@ func (h *Host) ConvertNMHostToAPI() *ApiHost {
 	a.Version = h.Version
 	a.IsDefault = h.IsDefault
 	a.NatType = h.NatType
+	a.PersistentKeepalive = int64(h.PersistentKeepalive)
```
Comment (Aceix) [inline models/api_host.go:62]: divide by `time.Second` for the UI

Comment (gabrielseibel1) [reply models/api_host.go]: https://github.com/gravitl/netmaker/pull/2604/commits/4828648a8a07d4f0bfd5ce85c56c0e3afc3bfe98

File: models/api_host.go (hunk 7 above)
Comment (gabrielseibel1) [reply models/api_host.go]: https://github.com/gravitl/netmaker/pull/2604/commits/4828648a8a07d4f0bfd5ce85c56c0e3afc3bfe98

**Review by theguy951357 — APPROVED** [review]

**Review by Aceix — APPROVED** [review]

**Review by abhishek9686 — APPROVED** [review]

**Pull request merged**